lib/model/message_list.dart

```diff
@@ -155,6 +155,27 @@ mixin _MessageSequence {
     _processMessage(messages.length - 1);
   }
 
+  /// Given message ids, remove the corresponding messages from the view, and
+  /// return true if at least one message has been removed, or false otherwise.
+  ///
+  /// If none of the message ids provided can be found, this is essentially
+  /// a no-op.
+  bool _removeMessagesById(Iterable<int> messageIds) {
+    bool hasRemovedMessage = false;
+    for (final messageId in messageIds) {
+      final messageIndex = _findMessageWithId(messageId);
+      if (messageIndex == -1) continue;
+
+      assert(contents.length == messages.length);
+      messages.removeAt(messageIndex);
+      contents.removeAt(messageIndex);
+      assert(contents.length == messages.length);
+      hasRemovedMessage = true;
+    }
+    if (hasRemovedMessage) _reprocessAll();
+    return hasRemovedMessage;
+  }
```

Comment:
One thing I notice in this implementation is that the `List.removeAt` calls can be expensive, if `messageIndex` is low and the list's length is large. From the [doc](https://api.flutter.dev/flutter/dart-core/List/removeAt.html); this part in particular:

> […] and moves all later objects down by one position.

There are ways we could be more efficient in that scenario (low `messageIndex`, large length), but I think they'd all come with tradeoffs in other scenarios. Also, bulk-delete events probably aren't very common anyway. But mentioning this in case you or Greg see something that's obviously an improvement.

Reply:
Yeah, the cost of this loop is quadratic, which is enough to make me nervous — with $n$ messages in the list, if the event has $k$ messages to delete and $d$ of them are in this message list, it's $\Theta(k \log n + d n)$. (The first term is from the `_findMessageWithId` lookups.)

In particular if you're looking at a conversation of $n$ messages, and an admin decides to delete the whole conversation so $d = k = n$, that's $\Theta(n^2)$.

An algorithm that would fix that is:
* First do the lookups, to determine if there are any affected messages in this list at all. That costs $\Theta(k \log n)$. (And saves us from spending $\Theta(n)$ on the next step if $d = 0$. If $d > 1$, we're resigned to spending that much time anyway.)
* Then, if there are any affected messages, use `List.removeWhere` to remove them from each of the lists all at once. (Build a `Set` out of `messageIds` first, to make that efficient.) This costs $\Theta(k + n)$. (The first term is for building the `Set`; the second for the `removeWhere` calls.)

That makes the total cost $O(k \log n + n)$ — no more quadratics. And it keeps it $O(k \log n)$ for the case where no messages were affected.

Reply:
Thanks for the time complexity analysis! So [`List.removeWhere`](https://api.dart.dev/stable/2.9.3/dart-core/List/removeWhere.html) performs in linear time? I couldn't find reference for how it is implemented. My guess is that they just make a copy of the list with the items removed.

Reply:
Looks like removing content with `List.removeWhere` wouldn't be quite straightforward because it is a list parallel to messages that don't have the message ID.

Reply:
> So [`List.removeWhere`](https://api.dart.dev/stable/2.9.3/dart-core/List/removeWhere.html) performs in linear time? I couldn't find reference for how it is implemented. My guess is that they just make a copy of the list with the items removed.

My guess is it's a bit more efficient than that in that it reuses the existing array, and just moves items down to lower indexes to cover gaps. But yeah, should definitely be linear time.

The docs on List are a bit lacking in spelling out the performance characteristics, I think in part because there isn't a great place to put that information; `List` itself is actually an interface, and implementations can have different performance characteristics. But on `List` it does say:
> The default growable list, as created by [], keeps an internal buffer, and grows that buffer when necessary. This guarantees that a sequence of [add](https://api.dart.dev/stable/3.4.4/dart-core/List/add.html) operations will each execute in amortized constant time. Setting the length directly may take time proportional to the new length, and may change the internal capacity so that a following add operation will need to immediately increase the buffer capacity. Other list implementations may have different performance behavior.

So basically for each method you should think of it as doing whatever is the natural most efficient thing it could be doing for that method, given that data structure.

That data structure is essentially the same as a Python list or a JS array, if that helps in thinking about it.

(btw that's an old Dart version in that docs link; if you delete the "2.9.3/" for the version so it just says "stable", it'll give you the current version.)

Reply:
> Looks like removing content with `List.removeWhere` wouldn't be quite straightforward because it is a list parallel to messages that don't have the message ID.

Hmm, I see.

In that case probably the best thing is to write the loop ourselves. Essentially do this:

> reuses the existing array, and just moves items down to lower indexes to cover gaps.

but do the testing only on `messages`, and the moving on both `messages` and `content`.